README.md
# Clojure Rest MySQL application

Clojure App to serve a MySQL database through RESTful APIs.

## Clojure libraries used

- [Ring](https://github.com/ring-clojure/ring) - HTTP Webserver

```
[ring "1.8.2"]
```

- [Compojure](https://github.com/weavejester/compojure) - Handle RESTful routing for Ring - Handles URL path, and method switching

```
[compojure "1.6.2"]
```

- [Clojure java.jdbc](https://github.com/clojure/java.jdbc) - A low-level Clojure wrapper for JDBC-based access to databases

```
[org.clojure/java.jdbc "0.7.11"]
```

- [next-jdbc](https://github.com/seancorfield/next-jdbc) - The next generation of clojure.java.jdbc: a new low-level Clojure wrapper for JDBC-based access to databases.

```
[seancorfield/next.jdbc "1.1.610"]
```

- [Clojure Data](https://github.com/clojure/data.json) - Clojure's JSON library - convert data types - to & from JSON

```
[org.clojure/data.json "1.0.0"]
```

## Usage

- Download the [data](https://istd50043.github.io/project) (.csv to be loaded into a mysql db) and save under /resources folder
- Add the library dependencies into `project.clj` file - should already be there
- `lein run 8000` to run at port 8000
- Test the REST APIs

## License

Copyright © 2020 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-10-20
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-10-20
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clojure-rest-mysql/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clojure-rest-mysql/compare/0.1.0...0.1.1


test/clojure_rest_mysql/core_test.clj
(ns clojure-rest-mysql.core-test
  (:require [clojure.test :refer :all]
            [clojure-rest-mysql.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/clojure_rest_mysql/core.clj
(ns clojure-rest-mysql.core
  (:require [clojure-rest-mysql.reviews.model :as reviews]
            [clojure-rest-mysql.reviews.handler :refer [handle-all-reviews
                                                        handle-create-review]])
  (:require [ring.adapter.jetty :as jetty]
            [ring.middleware.reload :refer [wrap-reload]]
            [ring.middleware.params :refer [wrap-params]]
            [ring.middleware.resource :refer [wrap-resource]]
            [ring.middleware.file-info :refer [wrap-file-info]]
            [compojure.core :refer [defroutes ANY GET POST PUT DELETE]]
            [compojure.route :refer [not-found]]
            [ring.handler.dump :refer [handle-dump]]
            [next.jdbc :as jdbc]))

; Initialise database -> datasource
(def db {:dbtype "mysql" :dbname "kindle-reviews" :user "root" :password "1234" :serverTimezone "UTC"})
(def ds (jdbc/get-datasource db))


(defn greet [req]
  {:status 200
   :body "Hello world"
   :headers {}})
(defn yo [req]
  (let [name (get-in req [:route-params :name])]
    {:status 200
     :body (str "Yo " name "!")
     :headers {}}))

; REST routes
(defroutes routes
  (GET "/" [] greet)
  (GET "/request" [] handle-dump)
  (GET "/yo/:name" [] yo)

  (GET "/reviews" [] handle-all-reviews)
  (POST "/reviews" [] handle-create-review)

  (not-found "Page not found"))

(defn wrap-db [hdlr]
  (fn [req]
    (hdlr (assoc req :clojure-rest-mysql/ds ds))))

(defn wrap-server [hdlr]
  (fn [req]
    (assoc-in (hdlr req) [:headers "Server"] "Kindle Reviews MySQL Server")))

(def app
  (wrap-server (wrap-db (wrap-params routes))))

(defn -main [port]
  (reviews/create-table ds)
  (jetty/run-jetty app {:port (Integer. port)}))

(defn -dev-main [port]
  (reviews/create-table ds)
  (jetty/run-jetty (wrap-reload #'app) {:port (Integer. port)}))

src/clojure_rest_mysql/reviews/handler.clj
(ns clojure-rest-mysql.reviews.handler
  (:require [clojure-rest-mysql.reviews.model :refer [read-all-reviews
                                                      create-review
                                                      read-a-review]]
            [ring.util.response :refer [response]]
            [clojure.data.json :as json]))


; asin char (10)
; helpful varchar (10)
; overall integer (1)
; reviewText text (1000)
; reviewTime varchar (11)
; reviewerID varchar (14)
; reviewerName varchar (64)
; summary varchar (255)
; unixReviewTime char (10)


(defn handle-all-reviews [req]
  (let [ds (:clojure-rest-mysql/ds req)
        reviews (read-all-reviews ds)]
    {:status 200
     :headers {}
    ;  :body (str reviews)}))
     :body (list (json/write-str (response reviews)))}))

(defn handle-a-review [req]
  (let [ds (:clojure-rest-mysql/ds req)
        asin (get-in req [:params "asin"])
        reviewerID (get-in req [:params "reviewerID"])
        unixReviewTime (get-in req [:params "unixReviewTime"])
        selected-review (read-a-review ds asin reviewerID unixReviewTime)]
    {:status 200
     :headers {}
    ;  :body (str reviews)}))
     :body (list (json/write-str (response selected-review)))}))


(defn handle-create-review [req]
  (let [ds (:clojure-rest-mysql/ds req)
        asin (get-in req [:params "asin"])
        helpful (get-in req [:params "helpful"])
        overall (get-in req [:params "overall"])
        reviewText (get-in req [:params "reviewText"])
        reviewTime (get-in req [:params "reviewTime"])
        reviewerID (get-in req [:params "reviewerID"])
        reviewerName (get-in req [:params "reviewerName"])
        summary (get-in req [:params "summary"])
        unixReviewTime (get-in req [:params "unixReviewTime"])
        ; review-asin (create-review ds asin, helpful overall reviewText reviewTime reviewerID reviewerName summary unixReviewTime)
        ]
    (create-review ds asin helpful overall reviewText reviewTime reviewerID reviewerName summary unixReviewTime)
    {:status 200
    ;  :headers {"Location" "/reviews"} ;redirect back to /reviews
    ;  :body (str "Successfully inserted a review into " review-asin)
     :body (str "Successfully inserted a review on book asin" "(" asin ")" " from reviewerID" "(" reviewerID ")")}))



src/clojure_rest_mysql/reviews/model.clj
(ns clojure-rest-mysql.reviews.model
  (:require [next.jdbc :as jdbc]))

; Kindle Reviews Dataset columns
; asin - ID of the product, eg B00FA64PK
; helpful - helpfulness rating of the review, eg 2/3 -- note that its raw format in csv is [2,3] - need to format
; overall - rating of the product, eg 3
; reviewText - text of the review (heading), can be very long
; reviewTime - time of the review (raw) - eg. 01 6, 2014 --need to convert later like this SELECT STR_TO_DATE ("01 5,2017", "%m %d, %Y");
; reviewerID - ID of the reviewer, eg A3SPTOKDG7WBLN
; reviewerName - name of the reviewer, eg ubavka seirovska
; summary - summary of the review (description) eg Not bad, not exceptional
; unixReviewTime - unix timestamp, eg 1388966400 --need to convert later like this mysql> SELECT FROM_UNIXTIME(1447430881); -> '2015-11-13 10:08:01'
;;; To check data types: https://www.w3schools.com/sql/sql_datatypes.asp

; CRUD OPERATIONS for kindle_reviews mysql database
; - READ all reviews
; - READ reviews of a specified book
; - CREATE a review for a specified book
; - Update the (written) review for a specified book
; - DELETE a (written) review for a specified book - user can only delete its own review

(def db {:dbtype "mysql" :dbname "kindle-reviews"})
(def ds (jdbc/get-datasource db))


; Load data into mysql from csv


; Insert sample data
(defn load-sample-data [ds]
  (jdbc/execute! ds ["INSERT INTO reviews (asin, helpful, overall, reviewText, reviewTime, reviewerID, reviewerName, summary, unixReviewTime)
                      VALUES (?,?,?,?,?,?,?,?,?), (?,?,?,?,?,?,?,?,?), (?,?,?,?,?,?,?,?,?)"
                     "B000F83SZQ" "[0, 0]" "5" "I enjoy vintage books and movies" "04 21, 2014" "A1F6404F1VG29J" "jack brown" "I enjoy vintage books and movies summary" "1398038400"
                     "B000FDJ0FS" "[2, 9]" "3" "I love this grape" "02 16, 2014" "A1CQ8WG6CUDBNV" "danny cake" "I love this grape summary" "1392508800"
                     "B000GFK7L6" "[3, 4]" "3" "cool stuff" "11 3, 2009" "A3GXR6CHHPX0JS" "Zach green" "cool stuff summary" "1257206400"]))

; Create 'reviews' Table
(defn create-table [ds]
  (jdbc/execute! ds ["
                      CREATE TABLE IF NOT EXISTS reviews (
                      asin char(10),
                      helpful varchar(10),
                      overall integer(1),
                      reviewText text(1000),
                      reviewTime varchar(11),
                      reviewerID varchar(14),
                      reviewerName varchar(64),
                      summary varchar(255),
                      unixReviewTime integer(10)
                      )"])
  (load-sample-data ds))





; Insert
(defn create-review [ds asin helpful overall reviewText reviewTime reviewerID reviewerName summary unixReviewTime]
  (jdbc/execute!
   ds
   ["INSERT INTO reviews (asin, helpful, overall, reviewText, reviewTime, reviewerID, reviewerName, summary, unixReviewTime)
                VALUES (?,?,?,?,?,?,?,?,?)"
    asin helpful overall reviewText reviewTime reviewerID reviewerName summary unixReviewTime]))



; Read All - wont really be used. Just for testing purpose
(defn read-all-reviews [ds]
  (jdbc/execute! ds ["SELECT * from reviews"]))
